---
title: "SH_scRNAseq"
author: "Michael Nehls"
date: "08/19/2020"
output: 
  html_document:
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
library(Seurat)
library(SeuratWrappers)
library(monocle3)
library(patchwork)
library(tidyverse)
library(cowplot)
library(stringr)
# sessionInfo()
```

## Importing the data and integrating the two Naive Data Sets

Using Seurat, data is imported from each 10X run. 
Then anchor genes are identified between the KO and WT conditions, and the data sets are integrated into 1. 
From here, PCA can be performed and UMAP plots can be generated. 
I imported the "filtered" matrices from the CellRanger output provided by the genome institute. 
It seems like they have already undergone some quality control. 
We also have the "raw" matrices and the initial FASTQ files. 
The packaged Mock treated files are not unpacking nicely (they are corrupted) on the cluster, 
so I still need to verify whether they are alright on our RAID system. 

```{r Import}
# loading Naive condition datasets for BMDMs from WT and KO genotypes
# setting them as Seurat Objects

renew_naive_integrated <- function() {
  wt_naive.data <- Read10X(data.dir = "WT_Naive/filtered_feature_bc_matrix/")
  wt_naive <- CreateSeuratObject(
                                 counts = wt_naive.data,
                                 project = "WT_Naive",
                                 min.cells = 3,
                                 min.features = 200
  )
  wt_naive@meta.data[, "condition"] <- "wt_naive"
  wt_naive
  
  ko_naive.data <- Read10X(data.dir = "KO_Naive/filtered_feature_bc_matrix/")
  ko_naive <- CreateSeuratObject(
                                 counts = ko_naive.data,
                                 project = "WT_Naive",
                                 min.cells = 3,
                                 min.features = 200
  )
  
  ko_naive@meta.data[, "condition"] <- "ko_naive"
  ko_naive
  
  #adding both naive objects into a list
  naive.list <- c(wt_naive, ko_naive)
  
  # before finding anchors between datasets, 
  # do standard preprocessing and identify variable features
  for (i in seq_along(naive.list)) {
    naive.list[[i]] <- NormalizeData(naive.list[[i]], verbose = FALSE)
    naive.list[[i]] <- FindVariableFeatures(
                                      naive.list[[i]],
                                      selection.method = "vst",
                                      nfeatures = 2000,
                                      verbose = FALSE
    )
  }  # vst = variance stabilizing transformation
  
  # identify anchors using FindIntegrationAnchors function,
  # takes list of Seurat objects as input
  
  naive.anchors <- FindIntegrationAnchors(object.list = naive.list, dims = 1:30)
  naive.integrated <- IntegrateData(anchorset = naive.anchors, dims = 1:30)
   
  # switch to integrated assay. The variable features of this assay are 
  # automatically set during IntegrateData
  
  DefaultAssay(naive.integrated) <- "integrated"
  
  # Run the standard workflow for visualization and clustering
  naive.integrated <- ScaleData(naive.integrated, verbose = FALSE)
  naive.integrated <- RunPCA(naive.integrated, npcs = 30, verbose = FALSE)
  naive.integrated <- RunUMAP(naive.integrated, reduction = "pca", dims = 1:30)
  saveRDS(naive.integrated, "files/naive.integrated.rds") 

  return(naive.integrated)
}

# Checks for an rds or renews the markers frame - set force = TRUE to renew all markers frames
get_naive_integrated <- function(force = FALSE) {
  if (!force & file.exists("files/naive.integrated.rds")){
    naive.integrated <- readRDS("files/naive.integrated.rds") %>% return()
  } else {
    naive.integrated <- renew_naive_integrated() %>% return()
  }
}
naive.integrated <- get_naive_integrated()

p1 <- DimPlot(
              naive.integrated,
              reduction = "umap",
              group.by = "condition",
              label = TRUE,
              repel = TRUE
)
p1
```

## We can also identify clusters via Seurat!

### Pooled Cluster Plot

```{r Clustering_merged}
renew_naive_integrated_UMAP <- function() {
  naive.integrated <- FindNeighbors(
                                  naive.integrated,
                                  reduction = "pca",
                                  dims = 1:30
  )
  naive.integrated <- FindClusters(naive.integrated, resolution = 0.5) %>% return()
  saveRDS(naive.integrated, "files/naive.integrated.UMAP.rds")
}
get_naive_integrated_UMAP <- function(force = FALSE) {
  if(!force & file.exists("files/naive.integrated.UMAP.rds")){
    naive.integrated <- readRDS("files/naive.integrated.UMAP.rds") %>% return()
  } else {
    naive.integrated <- renew_naive_integrated_UMAP() %>% return()
  }
}
naive.integrated <- get_naive_integrated_UMAP()

p2 <- DimPlot(naive.integrated, reduction = "umap", label = TRUE)
cowplot::plot_grid(p1, p2)
p2
```

### Cluster by condition
```{r Clustering_by_condition}
DimPlot(naive.integrated, reduction = "umap", split.by = "condition")

```

## Now we can plot the data and identify the gene expression in each of the clusters!

There are two primary ways to look at gene expression that I have found. 

### Looking per cluster at the highest expressed genes

The first is to ask what genes are most highly expressed in each cluster. 
This takes a while but gives a ton of data about which genes are determining cluster identity. 

I played around with setting cut-offs to sort the data similar to the Artyomov's shiny app. 

The last table generated by this code called "naive.markers" seems to give a summary of what genes
are most responsible for determining the identity of each cluster. 
The command that generates this can be fed parameters for only looking genes that are 
positively correlated with a cluster identity, defining a minimum percentage expression within the cluster,
and defining a minimum log-fold-change over other clusters to give more or less restrictive results.
My list was generated to only identify genes that are positively correlated with a cluster identity,
are expressed in a minimum of 25 percent of the cells in the cluster, and are at least expressed 0.25 
log-fold-change higher within the cluster. 

```{r FindConservedMarkers}
# Try to identify the cell type identities of the clusters
DefaultAssay(naive.integrated) <- "RNA"

# A function to get a dataframe of markers if one doesn't exist, or renew it if it does
# Also saves dataframe as both an rds and csv
renew_markers <- function(cluster_number) {
  markers <- FindConservedMarkers(
    naive.integrated,
    ident.1 = cluster_number,
    grouping.var = "condition",
    verbose = FALSE
  )
  write_rds(markers, str_c("./files/markers_rds/c", as.character(cluster_number), "markers.rds")) 
  write.csv(markers, 
    file = str_c("./files/csv_xlsx_files/c", as.character(cluster_number), "markers.csv"), 
    append = FALSE, row.names = TRUE)
  return(markers)
}

# Checks for an rds or renews the markers frame - set force = TRUE to renew all markers frames
get_markers <- function(cluster_number, force = FALSE) {
  markers_name = str_c("c", as.character(cluster_number), "markers")
  markers_file = str_c("./files/markers_rds/", markers_name, ".rds")
  if (!force & file.exists(markers_file)){
    markers <- read_rds(markers_file)
    return(markers)
  } else {
    markers <- renew_markers(cluster_number)
    return(markers)
  }
}


for (cluster in seq_along(levels(naive.integrated))) {
  markers <- get_markers(cluster - 1) 
  assign(str_c("c", as.character(cluster - 1), "markers"), markers, envir = .GlobalEnv)
}

c0markers_filtered <- c0markers[c0markers$wt_naive_pct.1 > 0.7 & c0markers$wt_naive_pct.2 < 0.7,]
head(c0markers_filtered)

c1markers_filtered <- c1markers[c1markers$wt_naive_pct.1 > 0.7 & c1markers$wt_naive_pct.2 < 0.6, ]
head(c1markers_filtered)

c2markers_filtered <- c2markers[c2markers$wt_naive_pct.1 > 0.7 & c2markers$wt_naive_pct.2 < 0.6, ]
head(c2markers_filtered)

## Compare clusters to identify the most significant genes in determining their identity
renew_naive_markers = FALSE
if (!file.exists("./files/markers_rds/naive_markers.rds") | renew_naive_markers == TRUE) {
  naive.markers <- FindAllMarkers(naive.integrated, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
  write_rds(naive.markers, "./files/markers_rds/naive_markers.rds")
  write.csv(naive.markers, "./files/csv_xlsx_files/naive_markers.csv", row.names = TRUE, append = FALSE)
} else {
  naive.markers <- read_rds("./files/markers_rds/naive_markers.rds")
}
naive.markers %>% group_by(cluster) %>% top_n(n = 4, wt = avg_logFC) # top 4 cluster defining genes for each cluster
```

## Comparing relative expression between closely related clusters

A similar method can be used to compare two clusters directly.
This can be helpful for determining identities of closely related clusters.

```{r cluster_comparisons}

# a function to generate the comparison table and save it to a csv 
comparison_markers <- function(cluster1, cluster2) {
  markers <- FindMarkers(naive.integrated, ident.1 = cluster1, ident.2 = cluster2, min.pct = 0.25) 
  write_rds(markers, str_c("./files/markers_rds/c", as.character(cluster1), "v", as.character(cluster2), "markers.rds")) 
  write.csv(markers, 
    file = str_c("./files/csv_xlsx_files/c", as.character(cluster1), "v", as.character(cluster2), "markers.csv"), 
    append = FALSE, row.names = TRUE)
  return(markers)
}

# Checks for an rds or renews the markers frame - set force = TRUE to renew all markers frames
get_comparison_markers <- function(cluster1, cluster2, force = FALSE) {
  markers_name = str_c("c", as.character(cluster1), "v", as.character(cluster2), "markers")
  markers_file = str_c("./files/markers_rds/", markers_name, ".rds")
  if (!force & file.exists(markers_file)){
    markers <- read_rds(markers_file)
    return(markers)
  } else {
    markers <- comparison_markers(cluster1, cluster2)
    return(markers)
  }
}

c2v3markers <- get_comparison_markers(2, 3)
c0v1markers <- get_comparison_markers(0, 1)
c5v8markers <- get_comparison_markers(5, 8)
c1v6markers <- get_comparison_markers(1, 6)
```


### Looking at expression of predetermined genes

The second is by asking what the expression profile is of makers that we know are associated with different cell types.

```{r featurePlot}
## Plot gene expression by feature - way faster!
p_macs <- FeaturePlot(naive.integrated, features = c("Cd14", "Cd68", "Cd62l", "Cd31", "Cx3cr1", "Ace", "Ccr2", "Lyz2", "Lpl"), min.cutoff = "q9")
p_macs
p_PMN <- FeaturePlot(naive.integrated, features = c("Clec4d", "S100a8", "S100a9", "Acod1"), min.cutoff = "q9")
p_PMN
p_DC <- FeaturePlot(naive.integrated, features = c("Ctsh", "Ptms", "Napsa", "Ccr3", "Itgae","Flt3","Cxcl16", "Syngr2"), min.cutoff = "q9")
p_DC
```

## Differential expression analysis

Now it is time for differential expression analysis! I am thinking about approaching it in two ways: 

* A differential expression gene table

* A volcano style-plot

### Generating the differential expression tables

```{r generate_cluster_diffexpression}

naive.integrated$celltype.condition <- paste(Idents(naive.integrated), naive.integrated$condition, sep = "_")
naive.integrated$celltype <- Idents(naive.integrated)
Idents(naive.integrated) <- "celltype.condition" # changes levels in naive integrated to one per cluster per condition

# A function to generate the differential expression table, assign it to a global variable, and save a csv
cluster_differential <- function(cluster) {
  differential_name = str_c("c", as.character(cluster), "_differential")
  differential_file_name = str_c("./files/csv_xlsx_files/", differential_name, ".csv")
  differential <- FindMarkers(naive.integrated, ident.1 = str_c(as.character(cluster), "_wt_naive"), 
                               ident.2 = str_c(as.character(cluster), "_ko_naive"), verbose = FALSE)
  write.csv(differential, file = differential_file_name, append = FALSE, row.names = TRUE)
  assign(differential_name, differential, envir = .GlobalEnv)
}

for (cluster_plus_one in seq_along(levels(naive.integrated$seurat_clusters))) {
  cluster_differential(cluster_plus_one - 1)
}
```

### Generating volcano-style plots

```{r Plot_Function_for_Diffex}
# A function to plot the differential expression in a volcano style plot with genes labeled
plot_cluster <- function(cluster, genes = c("Bhlhe40")) {
    Idents(naive.integrated) <- "celltype"
    cluster_subset <- subset(naive.integrated, idents = as.character(cluster))
    Idents(cluster_subset) <- "condition"
    avg_exp <- log1p(AverageExpression(cluster_subset, verbose = FALSE)$RNA)
    avg_exp$gene <- rownames(avg_exp)

    cluster_plot <- ggplot(avg_exp, aes(wt_naive, ko_naive)) +
        geom_point() +
        ggtitle(paste("cluster", as.character(cluster), sep = " "))
    cluster_plot <- LabelPoints(
                                cluster_plot,
                                points = genes,
                                repel = TRUE
    )
    cluster_plot
}

```

### Cluster 0

```{r cluster0_diffex_plot}
# Look at the beginning of the differential expression file to choose genes to label
head(c0_differential, 10)
genes_to_label <- c("Bhlhe40", "Rps3a1", "Rps7", "Rp19")

p_c0 <- plot_cluster(0, genes_to_label)
plot_grid(p2, p_c0) # p2 is included to give reference to which cluster we are looking into
```


### Cluster 1

```{r cluster1_diffexpression}
head(c1_differential, 10)
genes_to_label <- c("Apoe", "Mmp12", "Dhrs3", "Rp19")
p_c1 <- plot_cluster(1, genes_to_label)
plot_grid(p2, p_c1)
```


### Cluster 2

```{r cluster2_diffexpression}
head(c2_differential, n = 10)
genes_to_label <- c("Bhlhe40", "AA467197", "Ccl2", "Vim", "S100a9")

p_c2 <- plot_cluster(2, genes_to_label)
plot_grid(p2, p_c2)
```


### Cluster 3

```{r cluster3_diffexpression}
head(c3_differential, n = 10)
genes_to_label <- c("Ccl2", "Rps9", "Rps27a", "Rbpms", "Rpl5")

p_c3 <- plot_cluster(3, genes_to_label)
plot_grid(p2, p_c3)
```


### Cluster 4

```{r cluster4_diffexpression}
head(c4_differential, n = 10)
genes_to_label <- c("Ccl2", "AA467197", "Mmp12", "Rpl32", "Rps7")

p_c4 <- plot_cluster(4, genes_to_label)
plot_grid(p2, p_c4)
```

### Cluster 5

```{r cluster5_diffexpression}
head(c5_differential, n = 10)
genes_to_label <- c("Hpgds", "Lmna", "Bhlhe40", "Angptl2", "Crip1")

p_c5 <- plot_cluster(5)
plot_grid(p2, p_c5)

```

### Cluster 6

```{r cluster6_diffexpression}
head(c6_differential, n = 10)
genes_to_label <- c("Mmp12", "Apoe", "Lyz2", "Prdx5")

p_c6 <- plot_cluster(6, genes_to_label)
plot_grid(p2, p_c6)
```


### Cluster 7

```{r cluster7_diffexpression}
head(c7_differential, n = 10)
genes_to_label <- c("Bhlhe40", "S100a9", "Ccl2", "Clec4d", "Hpgds")

p_c7 <- plot_cluster(7, genes_to_label)
plot_grid(p2, p_c7)
```

### Cluster 8

```{r cluster8_diffexpression}
head(c8_differential, n = 10)
genes_to_label <- c("Sphk1", "AA467197", "Bhlhe40", "Mmp12")

p_c8 <- plot_cluster(8, genes_to_label)
plot_grid(p2, p_c8)
```


### Cluster 9

```{r cluster9_diffexpression}
head(c9_differential, n = 10)
genes_to_label <- c("Rbpms", "Nov", "S100a6", "Heatr1", "Trf")

p_c9 <- plot_cluster(9, genes_to_label)
plot_grid(p2, p_c9)
```


### Cluster 10

```{r cluster10_diffexpression}
head(c10_differential, n = 10)
genes_to_label <- c("Fth1", "Ftl1", "Lgals3", "Bhlhe40", "Vim")

p_c10 <- plot_cluster(10, genes_to_label)
plot_grid(p2, p_c10)
```


### Cluster 11

```{r cluster11_diffexpression}
head(c11_differential, n = 10)
genes_to_label <- c("Hpgds", "Bhlhe40", "Lmna", "AA467197", "Psap")

p_c11 <- plot_cluster(11, genes_to_label)
plot_grid(p2, p_c11)
```


### Cluster 12

```{r cluster12_diffexpression}
head(c12_differential, n = 10)
genes_to_label <- c("Mmp12", "Gpnmb", "Klhl24", "Adgre5", "Nov")

p_c12 <- plot_cluster(12, genes_to_label)
plot_grid(p2, p_c12)
```


### Cluster 13

```{r cluster13_diffexpression}
head(c13_differential, n = 10)
genes_to_label <- c("Bhlhe40", "Elp5", "Rab10", "Mocos", "Mmp12")

p_c13 <- plot_cluster(13, genes_to_label)
plot_grid(p2, p_c13)
```


### Cluster 14

```{r cluster14_diffexpression}
head(c14_differential, n = 10)
genes_to_label <- c("Rundc1", "Cebpb", "Lsm11", "Kdelr1", "Lgals1")

p_14 <- plot_cluster(14, genes_to_label)
plot_grid(p2, p_14)
```

## An attempt at pseudotime analysis

This requires the SeuratWrappers library and Monocle3.
It is pretty easy to generate!
I need to read/experiment more to know how to modify the way that Monocle3 clusters and partitions.
For example, I want to try to partition by what we think the cells are.
For now, I have used the cluster with maximal c-kit expression as the root (time 0) of the analysis.

```{r pseudotime_experimentation}
cds <- as.cell_data_set(naive.integrated)
cds <- cluster_cells(cds)
p_monocle_clusters <- plot_cells(cds, show_trajectory_graph = FALSE) + ggtitle("Monocle3's Clusters")
p_monocle_partitions <- plot_cells(cds, color_cells_by = "partition", show_trajectory_graph = FALSE) + ggtitle("Monocle3's Partitions")
wrap_plots(p_monocle_clusters, p_monocle_partitions) 
  

integrated.sub <- subset(as.Seurat(cds), monocle3_partitions == 1)
cds <- as.cell_data_set(integrated.sub)
cds <- learn_graph(cds)
plot_cells(cds, label_groups_by_cluster = FALSE, label_leaves = FALSE, label_branch_points = FALSE)

# Set max Kit expression as the root of our pseudotime tree
max.kit <- which.max(unlist(FetchData(integrated.sub, "Kit")))
max.kit <- colnames(integrated.sub)[max.kit]
cds <- order_cells(cds, root_cells = max.kit)
plot_cells(cds, color_cells_by = "pseudotime", label_cell_groups = FALSE, label_leaves = FALSE, 
           label_branch_points = TRUE) +
            ggtitle("Pseudotime map with max Kit expression as root")


# Set the assay back as 'integrated'
integrated.sub <- as.Seurat(cds, assay = "integrated")
p_pseudotime <- FeaturePlot(integrated.sub, "monocle3_pseudotime")
wrap_plots(p2, p_pseudotime)

```

Look at that! This is pretty exciting! I don't have any breakthrough interpretations, but it looks like clusters 2 and 5 have the highest expression of c-kit. It looks like cluster 6 and 13 are very mature, which fits with our interpretation of them as mature PMNs. That is all I really can say, given how early on I am in understanding this analysis. But just for fun, I will overinterpret it! It looks like there may be two distinct paths forward in pseudotime. The one to the left would fit with the interpretation that those clusters are DC like. The one to the right is fuzzier, expecially because given cell-types don't cluster as nicely. It looks like cluster 12 may be a less mature PMN cluster. And maybe cluster 9 is a mature macrophage like cluster (expresses Ccr2 highly)? It is a bit weird that clusters 0 and 1 have multiple pseudotimes attributed to them. I don't know what that means. I also need to better understand the significance of the branch points. Let me know if you have any questions! That will help guide what I learn about!  


```{r scVelo setup}
library(SeuratDisk)
renew_h5ad <- function() {
  DefaultAssay(naive.integrated) <- "RNA"
  SaveH5Seurat(naive.integrated, filename = "files/naive.integrated.h5Seurat")
  Convert("files/naive.integrated.h5Seurat", dest = "h5ad")
}
check_h5ad <- function(force = FALSE) {
  if(!forec & file.exists("files/naive.integrated.h5ad")){} else{renew_h5ad()}
}
```